### DISCUSSION TITLE
Calling Mistral API from edge function

### DISCUSSION THREAD

**Discussion by doagenailabs** [vercel/community#5361] [Feature Question]
### Question

I have several serverless functions plus an edge function on my vercel project. The are acting as middleware for integrating Twilio with Mistral (AI). The process is, based on events, Twilio calls a serverless function, api/index.js. This function calls another serverless function, that eventually has to call my edge function. This edge function is the one calling Mistral API. 

The edge function is edge type (and not standard serverless) because Mistral can take sometimes longer than 10 seconds to respond (standard timeout for serverless in hobby account). This is the function:

```

export const config = {
  runtime: 'edge',
};

export async function queryMistral(userMessage, systemMessage, mistralURL, mistralApiKey, model, temperature, max_tokens, previousMessages) {

    // Check if running on the Edge Runtime
    if (typeof globalThis.EdgeRuntime !== 'string') {
        console.log('queryMistral is running on the Edge Runtime: ', globalThis.EdgeRuntime);
    } else {
        console.log('queryMistral is not running on the Edge Runtime: ', globalThis.EdgeRuntime);
    }

    // Constructing the request body
    const requestBody = {
        model: model,
        messages: [{ role: "system", content: systemMessage }, ...previousMessages],
        temperature: temperature,
        max_tokens: max_tokens
    };

    try {
        const response = await fetch(`${mistralURL}`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'Authorization': `Bearer ${mistralApiKey}`
            },
            body: JSON.stringify(requestBody)
        });

        if (!response.ok) {
            // Handle HTTP errors
            throw new Error(`HTTP error! status: ${response.status}`);
        }

        console.log('Mistral chat completion request sent, awaiting response');
        const responseData = await response.json(); 

        if (responseData.error) {
            throw new Error(`Error in Mistral response: ${responseData.error}`);
        }

        return responseData; 
    } catch (error) {
        console.error(`Error during Mistral query setup: ${error.message}`);
        throw error;
    }
}

```

It only works sometimes. Most of the times I see on vercel's logs a 504 timeout. My guess is that it's not acting as an edge function, even though that is my intention. My package.json contains "type": "module". That together with "export const config = { runtime: 'edge',};" should be enough to activate it, right? 

I'm calling this function from another serverless function, like this:

```

                console.log("Using Mistral for response generation.");
                const mistralResponse = await queryMistral(userMessage, systemMessage, mistralURL, mistralApiKey, mistralModel, mistralModelTemperature, max_tokens, previousMessages);

                if (mistralResponse.error) {
                    throw new Error(mistralResponse.error);
                }

                // Extracting the generated text and additional data from the response
                const generatedText = mistralResponse.generated_text;
                if (!generatedText) {
                    throw new Error('No generated text received from Mistral');
                }

                return generatedText;
```

That internal call (from serverless to edge function) is valid? Or is it mandatory in this case to call it externally, via HTTP request? If so, what would be the URL and the token?

If edge function was really there, I shouldn't be getting any 504 timeouts from that function unless it doesn't reply back within the first 25 seconds, right?

This is what I see in vercel logs (truncated):

Request Path
/api
Status Code
504
Level
Error

Using Mistral for response generation.
queryMistral is running on the Edge Runtime:  undefined

**Comment by z4nr34l** [comment]
Maybe you can simplify this using [AI SDK](https://sdk.vercel.ai/docs)?

That would be much easier solution IMHO then creating everything from 0.

Here you will find docs related to Mistral model:

https://sdk.vercel.ai/docs/guides/providers/mistral
